design/0001-metallb-crd.md

```diff
@@ -0,0 +1,173 @@
+# MetalLB Custom Resource Controller 
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+<!-- /toc -->
+
+## Summary
+Today MetalLB allows only configuration of layer2 and BGP protocols using Configmap objects, 
+which is simple from configuration perspective and doesn't require cluster-admin rights,
+however it doesn't have validation on the API server level, no support for API versioning and 
+no fine-grained permission model via RBAC.
+
+## Motivation
+Custom resource definition "CRD" takes advantage of kubernetes extensibility to deliver the automation advantages of cloud services like 
+provisioning, scaling and backup/restore.
+
+### Goals
+Configuring MetalLB layer2 and BGP using kubernetes custom resource "CRD" instead of ConfigMap.
+
+## Proposal
+Using SIG's [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) framework to build new CRD "MetalLB" object 
+and a controller which works as a translator from the user's configuration yaml to a MetalLB's configmap object.
```

Comment:
Why would we continue to use the intermediate ConfigMap?

Reply:
the controller convert the CRD configs to configmap intentionally to avoid chaning Metallb provisiong logic with that approach we have no risk at all if users wanted to continue with configmap

Reply:
Looks like @russellb agrees and this should be an alternative configuration method.